WIP: Merge noise-info features from VGA internal branch

sasamiyatu opened the pull request:
This is a work in progress. Do not merge this yet.

Requesting review to check that the buffers output by the various noise-data options are what are expected. Some may have changed unintentionally, as I had to change the computations somewhat to fit the current internal structure of the path tracing loop.

Commit:
add bounce-data rendering

Commit:
some cleanup

Commit:
Fix g-buffer to not include raster

Commit:
fix options

sasamiyatu committed:
add material-id, metallic feature and roughness feature

Original commit:

Author: Saku Haikio <saku.haikio@tuni.fi>
Date:   Tue Nov 7 12:11:45 2023 +0200

    add material-id, metallic feature and roughness feature

sasamiyatu committed:
add pdf features

Original commit:

commit 3e3e786c09fe1dc5c178b47931051b88c01ad8a8
Author: Saku Haikio <saku.haikio@tuni.fi>
Date:   Thu Jan 25 11:34:23 2024 +0200

    add pdf features

sasamiyatu committed:
pdf-contribution renderer

sasamiyatu committed:
Adjust noise-data computation

sasamiyatu committed:
BMFR buffer stuff

makitalo reviewed (approved):
material-id should be moved from --noise-data to --renderer at some point, and it might be better not to rely on textures for the material index, but those are not critical, this could be merged already. All the features seem to work correctly.

sasamiyatu committed:
Make BMFR work with variable feature count

sasamiyatu committed:
Make bounce data work in BMFR

sasamiyatu committed:
Make noise-data stuff work again

sasamiyatu committed:
Change hard coded minmax buffer size to use actual number of normalized features

sasamiyatu committed:
Enable metallic, roughness and instance-id options in BMFR

Material and instance-id gbuffers are required only when the
options require them, and the code to read them is injected
with preprocessor macros. Therefore performance should be unaffected
when they are not used.

sasamiyatu commented:
I added metallic, roughness and instance-id options to BMFR, these should hopefully work correctly.

The --noise-data-str parameter should probably be separated to at least use a different name for BMFR, as the gbuffer options aren't related to the noise data renderer.

There seems to be some work in progress stuff in the noise data renderer from Saku to allow a maximum of 8 noise data features instead of 4 (using 2 buffers), I could have a look to see if this could be integrated with BMFR. I think some additional testing is needed to make sure the noise data stuff works correctly with BMFR (mainly that the buffers have the expected data in them when used with BMFR).

makitalo commented:
@sasamiyatu thanks, the BMFR output with metallic, roughness and instance-id looks reasonable, so it probably works correctly, but I didn't verify the contents of the buffers.

I agree that --noise-data-str should rather be for example --bmfr.buffers (similar to how SVGF and ReSTIR specify their own parameters).

If you can take a look at the integration of max 8 noise data features, that would be great.
